File: src/06.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ROWS 131
#define MAX_COLS 131
#define TRAJECTORY_START_SIZE 20

typedef struct {
    int x;
    int y;
    int direction;
} Position;

typedef struct {
    Position* data;
    int length;
    int _data_size;
} Trajectory;

void simulate_guard(char map_data[MAX_ROWS][MAX_COLS], int rows, int cols, Position start_pos, int obstacle_x, int obstacle_y, int visited[MAX_ROWS * MAX_COLS][4], int* is_loop, Trajectory* trajectory) {
    int x = start_pos.x, y = start_pos.y, direction = start_pos.direction;
    int flat_visited_size = rows * cols;

    // Directions: up, right, down, left
    static const int dx[] = {-1, 0, 1, 0};
    static const int dy[] = {0, 1, 0, -1};

    // Allocate initial trajectory array
    if (trajectory->length == 0) {
        trajectory->length = 1;
        trajectory->data = (Position*)malloc(TRAJECTORY_START_SIZE * sizeof(Position));
        trajectory->_data_size = TRAJECTORY_START_SIZE;
        trajectory->data[0] = start_pos;
    }

    for (int idx = 0; idx < trajectory->length; idx++) {
        Position elem = trajectory->data[idx];
        visited[elem.x * cols + elem.y][elem.direction] = 1;
    }

    while (1) {
        int nx = x + dx[direction];
        int ny = y + dy[direction];

        if (!(nx >= 0 && nx < rows && ny >= 0 && ny < cols)) {
            // Out of bounds, stop simulation
            break;
        }

        if ((map_data[nx][ny] == '#' || (nx == obstacle_x && ny == obstacle_y))) {
            // Turn right: increment direction by 1 and take modulo 4
            direction = (direction + 1) % 4;
        } else {
            // Move forward
            x = nx;
            y = ny;
            int flat_index_new = x * cols + y;

            if (visited[flat_index_new][direction]) {
                // Loop detected, stop simulation
                if (is_loop != NULL) {
                    *is_loop = 1;
                }
                break;
            } else {
                visited[flat_index_new][direction] = 1;

                // Add new position to trajectory
                int idx = trajectory->length++;
                if (trajectory->length > trajectory->_data_size) {
                    trajectory->_data_size = trajectory->_data_size * 2;
                    trajectory->data = (Position*)realloc(trajectory->data, trajectory->_data_size * sizeof(Position));
                }
                trajectory->data[idx].x = x;
                trajectory->data[idx].y = y;
                trajectory->data[idx].direction = direction;
            }
        }
    }
}

int main() {
    char map_data[MAX_ROWS][MAX_COLS];
    int rows = 0, cols = 0;

    char *buffer = NULL;
    size_t len = 0;
    while ((getline(&buffer, &len, stdin)) != -1) {
        if (cols == 0) {
            for (int i = 0; buffer[i] != '\n'; i++) {
                cols++;
            }
        }
        strncpy(map_data[rows], buffer, cols);
        rows++;
    }

    // Find guard's initial position and direction
    Position start_pos;
    int found_guard = 0;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (map_data[i][j] == '^') {
                start_pos.x = i;
                start_pos.y = j;
                start_pos.direction = 0;
                found_guard = 1;
                break;
            }
        }
        if (found_guard) {
            break;
        }
    }

    if (!found_guard) {
        printf("No guard found!\n");
        return 1;
    }

    // Find all reachable positions
    int visited[MAX_ROWS * MAX_COLS][4] = {0};
    Trajectory trajectory;
    trajectory.length = 0;
    trajectory._data_size = 0;
    trajectory.data = NULL;

    simulate_guard(map_data, rows, cols, start_pos, -1, -1, visited, NULL, &trajectory);

    // Find the number of reachable positions
    int reachable_positions_count = 0;
    int trajectory_length = trajectory.length;
    Position* reachable_positions = (Position*)malloc(trajectory_length * sizeof(Position));
    int* trajectory_lengths = (int*)malloc(trajectory_length * sizeof(int));
    memset(visited, 0, sizeof(visited));
    for (int i = 0; i < trajectory_length; i++) {
        Position current_pose = trajectory.data[i];
        int flat_index = current_pose.x * cols + current_pose.y;

        if (!visited[flat_index][0]) {
            visited[flat_index][0] = 1;
            reachable_positions[reachable_positions_count] = current_pose;
            trajectory_lengths[reachable_positions_count] = i;
            reachable_positions_count++;
        }
    }

    printf("Number of reachable positions: %d\n", reachable_positions_count);

    // Count loop-inducing positions
    int loop_count = 0;
    for (int i = reachable_positions_count-1; i > 0; i--) {
        int is_loop = 0;
        trajectory.length = trajectory_lengths[i];
        memset(visited, 0, sizeof(visited));
        simulate_guard(map_data, rows, cols, trajectory.data[trajectory_lengths[i]-1], reachable_positions[i].x, reachable_positions[i].y, visited, &is_loop, &trajectory);
        if (is_loop) {
            loop_count++;
        }
    }

    printf("Number of positions where obstacles would create loops: %d\n", loop_count);

    // Free allocated memory
    free(trajectory.data);
    free(reachable_positions);

    return 0;
}

File: src/06.py
import sys
from tqdm import tqdm


def simulate_guard(map_data, start_pos, obstacle_x, obstacle_y, trajectory=None):
    rows = len(map_data)
    cols = len(map_data[0])
    x, y, direction = start_pos
    
    if trajectory is None:
        trajectory = [(x, y, direction)]
    
    # Create a flattened index for the visited map
    flat_visited_size = rows * cols
    visited = [None] * flat_visited_size
    for (tx, ty, td) in trajectory:
        visited[tx * cols + ty] = [td]
    
    while True:
        # Directions: up, right, down, left
        if direction == 0:
            nx, ny = x - 1, y
        elif direction == 1:
            nx, ny = x, y + 1
        elif direction == 2:
            nx, ny = x + 1, y
        elif direction == 3:
            nx, ny = x, y - 1
        
        if not (0 <= nx < rows and 0 <= ny < cols):
            # Out of bounds, stop simulation
            break
        if map_data[nx][ny] == '#' or (nx, ny) == (obstacle_x, obstacle_y):
            # Turn right: increment direction by 1 and take modulo 4
            direction = (direction + 1) % 4
        else:
            # Move forward
            x, y = nx, ny
        
        trajectory.append((x, y, direction))
        flat_index_new = x * cols + y
        
        if visited[flat_index_new] is not None and direction in visited[flat_index_new]:
            # Loop detected, stop simulation
            return True, trajectory
        else:
            if visited[flat_index_new] is None:
                visited[flat_index_new] = []
            visited[flat_index_new].append(direction)
    
    return False, trajectory

def main():
    input_data = sys.stdin.read().strip()
    map_data = input_data.split('\n')
    
    # Find guard's initial position and direction
    start_pos = None
    for i, row in enumerate(map_data):
        for j, char in enumerate(row):
            if char == '^':
                start_pos = (i, j, 0)
                break
        if start_pos:
            break
    
    if not start_pos:
        print("No guard found!")
        return
    
    # Find all reachable positions
    _, trajectory = simulate_guard(map_data, start_pos, -1, -1)
    
    # Extract all unique (x, y) positions that were visited
    reachable_positions = []
    reachable_poses = []
    trajectory_lengths = []
    for idx, pose in enumerate(trajectory):
        if pose[:2] not in reachable_positions:
            reachable_positions.append(pose[:2])
            reachable_poses.append(pose)
            trajectory_lengths.append(idx)
    
    print("Number of reachable positions:", len(reachable_positions))
    
    # Count loop-inducing positions
    loop_count = 0
    for idx in tqdm(range(len(reachable_poses)-1, 0, -1), desc="Placing Obstacles"):
        obstacle_x, obstacle_y = reachable_positions[idx][:2]
        if (obstacle_x, obstacle_y) == start_pos[:2]:
            continue
        is_loop, _ = simulate_guard(map_data, trajectory[trajectory_lengths[idx]-1], obstacle_x, obstacle_y, trajectory[:trajectory_lengths[idx]])
        if is_loop:
            loop_count += 1
    
    print("Number of positions where obstacles would create loops:", loop_count)

if __name__ == "__main__":
    main()